### PR TITLE
docs: explain https support

### PR DESCRIPTION
**Pull request by JohannesMessner**
Still ToDo:

- [x] wait for tls to be implemented with all protocols: https://github.com/jina-ai/jina/issues/4492
- [x] align naming with this pr: https://github.com/jina-ai/jina/pull/4498
- [x] rename `https` -> `tls`, or whatever keyword the PRs above land on

### CODE REVIEW

**Comment by JohannesMessner** [comment]
Honestly I have no idea what I am talking about here so let me know if this makes any sense

File: docs/fundamentals/flow/flow-api.md [hunk 1]
```diff
@@ -238,6 +238,34 @@ f = Flow(protocol='http', expose_graphql_endpont=True)
...
+    protocol='http',
+    uvicorn_kwargs=[
+        f'ssl_certfile: {cert}',
+        f'ssl_keyfile: {key}',
+        f'ssl_keyfile_password: {pwd}',
+    ],
+)
+```
+
+HTTPS connection is supported in combination with the `http` and `websocket` protocols.
```
Comment (samsja) [inline docs/fundamentals/flow/flow-api.md:261]: it is weird to say https is supported with websockets, but I guess that this is more an issue with the naming of the tls feature

**Review by samsja — DISMISSED** [review]
lgtm

maybe you should wait to merge  that this [pr](https://github.com/jina-ai/jina/pull/4498) which will include the renaming is merged ?

**Comment by JoanFM** [comment]
@jacobowitz @deepankarm, please keep an eye on this one

File: docs/fundamentals/flow/flow-api.md [hunk 2]
```diff
@@ -238,6 +238,34 @@ f = Flow(protocol='http', expose_graphql_endpont=True)
 For more details about the Jina GraphQL enpoint, see {ref}`here <flow-graphql>`.
 ````
 
+(flow-https)=
+### Enable HTTPS
+
+You can enable HTTPS connectivity to your Flow by configuring the underlying [uvicorn](https://www.uvicorn.org/) server.
+To do this, you need an `ssl_certificate`, an `ssl_keyfile`, and an `ssl_keyfile_password`, which you pass to `uvicorn_kwargs`
```
Comment (jacobowitz) [inline docs/fundamentals/flow/flow-api.md:245]: You are mentioning uvicorn below but I would point out here as well that the cert/key/password variables have the excact same meaning as in uvicorn. I think thats not obvious?

**Review by jacobowitz — CHANGES_REQUESTED** [review]
Just one comment, rest looks good :)
And I think we should wait for the TODOs you mentioned already.

File: docs/fundamentals/flow/flow-api.md (hunk 2 above)
Comment (JohannesMessner) [reply docs/fundamentals/flow/flow-api.md:245]: Good point, will clarify

Comment (JoanFM) [reply docs/fundamentals/flow/flow-api.md:245]: maybe link to a documentation page on uvicorn explaining these parameters?

Comment (JohannesMessner) [reply docs/fundamentals/flow/flow-api.md:245]: > maybe link to a documentation page on uvicorn explaining these parameters?

It is linked at the bottom in the see also box, but I will move it to the top, needs to be more visible

**Comment by JoanFM** [comment]
What's the status here @JohannesMessner ?

**Comment by JohannesMessner** [comment]
> What's the status here @JohannesMessner ?

Blocked by implementing tls for all protocols (https://github.com/jina-ai/jina/issues/4492), as as that is merged this can be completed.

**Review by samsja — APPROVED** [review]

File: docs/fundamentals/flow/client.md [hunk 3]
```diff
@@ -3,103 +3,83 @@
...
     f.block()
 ```
 
-To connect a Client to the Flow's gateway, you need to specify an IP host address and port on which the Flow can be reached.
-This is done using the `host=` and `port=` constructor keywords.
-If the Client and the Flow gateway are running on the same machine, the `host=` parameter can be omitted, as it defaults to `'0.0.0.0'`.
+To connect to the `Flow`, the Client has to specify the followings parameters.
+All af these have to match the Flow and how it was set up:
+* the `protocol` it needs to use to communicate with the Flow
+* the `host` and the `port`on which the Flow is exposed
```
Comment (jacobowitz) [inline docs/fundamentals/flow/client.md:29]:
```suggestion
* the `host` and the `port` as exposed by the Flow
```

File: docs/fundamentals/flow/client.md [hunk 4]
```diff
@@ -108,16 +88,64 @@ Client(host='my.awesome.flow:1234', protocol='http', tls=True)
...
 Client(host='https://my.awesome.flow:1234', port=4321)
 ```
 ````
 
+````{admonition} Hint
+:class: hint
+The arguments above have usefule defaults: `protocol='grpc'` and `host='0.0.0.0'`.
+This is particularly useful when debugging or accessing a Flow on your local machine.
 
+To connect to a Flow `f` it is therefore often enought to do the following:
```
Comment (jacobowitz) [inline docs/fundamentals/flow/client.md:103]:
```suggestion
To connect to a Flow `f` it is therefore often enough to do the following:
```

File: docs/fundamentals/flow/client.md [hunk 5]
```diff
@@ -108,16 +88,64 @@ Client(host='my.awesome.flow:1234', protocol='http', tls=True)
...
+client.post('/endpoint', DocumentArray([d1, d2]))  # DocumentArray
+
+client.post('/endpoint')  # Empty
+```
+
+
+```{admonition} Caution
+:class: caution
+`Flow` also provides a `.post()` method that follows the same interface as `client.post()`.
+However, once your solution is deployed in the cloud, the Flow interface is not present anymore.
```
Comment (jacobowitz) [inline docs/fundamentals/flow/client.md:146]:
```suggestion
However, once your solution is deployed remotely, the Flow interface is not present anymore.
```

File: docs/fundamentals/flow/flow-api.md [hunk 6]
```diff
@@ -211,10 +211,46 @@ f = Flow(protocol='http', no_debug_endpoints=True, no_crud_endpoints=True)
...
-### Add GraphQL endpoint
+(cors)=
+### Enable Cross-Origin Resource Sharing (CORS)
 
-````{admonition} Attention
-:class: attention
+To make a Flow accessible from a website with a different domain, you need to enable [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).
+Among other things, CORS is necessary to provide a {ref}`Swagger UI interface <swagger-ui>` for your Flow.
+
+Not that CORS is disabled by default, for security reasons.
```
Comment (jacobowitz) [inline docs/fundamentals/flow/flow-api.md:220]:
```suggestion
Note that CORS is disabled by default, for security reasons.
```

File: docs/fundamentals/flow/flow-api.md [hunk 7]
```diff
@@ -211,10 +211,46 @@ f = Flow(protocol='http', no_debug_endpoints=True, no_crud_endpoints=True)
...
+```python
+from jina import Flow
+
+f = Flow(cors=True, protocol='http')
+```
+
+### Advanced configuration options
+
+HTTP support in Jina is powered by [Uvicorn](https://www.uvicorn.org/).
+You can configure the Flow's internal Uvicorn sever to your heart's content by passing `uvicorn_kwargs` to the Flow:
```
Comment (jacobowitz) [inline docs/fundamentals/flow/flow-api.md:231]: your heart's content?

**Review by jacobowitz — CHANGES_REQUESTED** [review]
Looks good, only some typos and small language things to fix I think

File: docs/fundamentals/flow/flow-api.md (hunk 7 above)
Comment (JohannesMessner) [reply docs/fundamentals/flow/flow-api.md:231]: Yeah, as much as you would like. Can use less flowery language if this is too much

**Review by jacobowitz — APPROVED** [review]
looks good to me :+1:

**Pull request merged**